### ISSUE TITLE
Parallelize `zplugin update --all`

### ISSUE THREAD

**Issue by gibfahn** [zdharma/zinit#184]
It would be great to run the module updates in parallel.

### Background

I have the following lines in my dotfiles setup/update script:

```bash
gitCloneOrUpdate zdharma/zplugin "$XDG_DATA_HOME/zsh/zplugin/bin" # Zsh plugin manager.
zsh -ic 'zplugin self-update; zplugin update --all' # Update zsh plugins.
```

This adds a fair amount of time to my update script, even with a modest number of dependencies and no files changed.

<details><summary>Raw text</summary>

```bash
···❯ time zsh -ic 'zplugin self-update; zplugin update --all' # Update zsh plugins.                                                                                                                                                                 2019-10-14 12:30:43 ~/code/dot (master)
Reusing gibrc...
Already up to date.
Current branch master is up to date.
Compiling Zplugin (zcompile)...
Note: updating also unloaded plugins

Updating plugin mafredri/zsh-async
Already up to date.
Current branch master is up to date.

Updating plugin trapd00r/LS_COLORS
Already up to date.
Current branch master is up to date.

Updating plugin zdharma/fast-syntax-highlighting
Already up to date.
Current branch master is up to date.

Updating plugin zsh-users/zsh-autosuggestions
Already up to date.
Current branch master is up to date.

Updating plugin zsh-users/zsh-completions
Already up to date.
Current branch master is up to date.
zsh -ic 'zplugin self-update; zplugin update --all'  0.98s user 1.07s system 13% cpu 15.358 total
```

</details>

<img width="891" alt="CopyQ usDSWp" src="https://user-images.githubusercontent.com/15943089/66749758-8f5e6c80-ee82-11e9-91b8-6b2115bdc6ee.png">

**Comment by psprint** [comment]
Yes it's true that this would be good enhancement. I was just thinking about "How to make the messages (e.g. new commits lists) survive in the parallel mode". However, I now see that the parallel update would be an option `-p/--parallel`, so user would be able to decide whether he wants to read the messages or not.

This is a little bigger thing, I will need to find some time for it.

**Comment by gibfahn** [comment]
>How to make the messages (e.g. new commits lists) survive in the parallel mode

Ideally it would be nice if each spawned thread stored its stdout in a buffer and then when they got joined up at the end the joiner could concat the stdout/err and take the worst of the return codes.

>This is a little bigger thing, I will need to find some time for it.

Yeah absolutely, it's both complex and a nice-to-have rather than a bug, so not a huge priority or very urgent. If I get the time I'll take a look at it myself.

**Comment by Seirdy** [comment]
Seconding this. It would be great if it could detect the number of CPU threads or
allow the user to specify a max number of threads.

Until this is implemented, I use the following commands to generate a script that
updates all my zplugin plugins in parallel:

``` sh
zplugin list | awk '{print $1}' | sed -e 's/^/zplugin update /g' -e 's,\x1B\[[0-9;]*[a-zA-Z],,g' -e 's/$/ \&/g' > "$HOME/.local/bin/zplg-update-parallel.zsh
chmod +x "$HOME/.local/bin/zplg-update-parallel.zsh
```

After running the above, typing `zplg-update-parallel.zsh` updates my plugins in
parallel. Unfortunately, it doesn't update stuff like snippets.

**Comment by psprint** [comment]
Could maybe a donation be granted for this? Like $5 or $10?

**Comment by gibfahn** [comment]
>Could maybe a donation be granted for this? Like $5 or $10?

Sure! Done.

**Comment by psprint** [comment]
Thanks! I'll start working this week.

**Comment by psprint** [comment]
I've finished implementing the first version of the feature. The code turned out to be quite nice and pleasant to do – hey `wait` is actually pretty usable tool :) The effect is this: https://asciinema.org/a/303174. You can pass `-q`/`--quiet` option to suppress the status messages.

**Comment by gibfahn** [comment]
Cool, looks like all I have to do is change:

```diff
-zsh -ic 'zplugin self-update; zplugin update --all' # Update zsh plugins.
+zsh -ic 'zplugin self-update; zplugin update --all -p 20' # Update zsh plugins.
```

To run 20 updates in parallel.

**Comment by gibfahn** [comment]
It's a bit odd that it prints `Spawning the next 20 concurrent update jobs...` twice though when I only have 11 plugins. Does it spawn more than 1 task per plugin?

![image](https://user-images.githubusercontent.com/15943089/75562920-52fe6800-5a41-11ea-8e6e-8366a368340d.png)

**Comment by psprint** [comment]
Snippets and plugins are being updated separately, hence the two groups. I'll implement a uniform update soon.
